File: stack.py
class Stack:
    def __init__(self):
        self._stk = []
        self._len = 0

    def push(self, x):
        if len(self._stk) > self._len:
            self._stk[self._len] = x
        else:
            self._stk.append(x)
        self._len += 1
        return self

    def pop(self):
        if self.is_empty():
            return None, IndexError("Stack.pop: empty stack")
        self._len -= 1
        x = self._stk[self._len]
        self._stk[self._len] = None
        return x, None

    def peek(self):
        if self.is_empty():
            return None, IndexError("Stack.peek: empty stack")
        return self._stk[self._len-1], None

    def is_empty(self):
        return self._len == 0

    def __str__(self):
        return "<" + ", ".join(str(x) for x in self._stk[:self._len]) + ">"

File: srv/js/stack.js
"use strict"

class Stack {
	constructor() {
		this._stk = []
		this._len = 0
	}

	push(x) {
		this._stk[this._len] = x
		this._len++
		return this
	}

	pop() {
		if (this.isEmpty())
			return [null, TypeError("Stack.pop: empty stack")]
		this._len--
		const x = this._stk[this._len]
		this._stk[this._len] = null
		return [x, null]
	}

	peek() {
		if (this.isEmpty())
			return [null, TypeError("Stack.peek: empty stack")]
		return [this._stk[this._len-1], null]
	}

	isEmpty() {
		return this._len === 0
	}

	toString() {
		return "<" + this._stk.slice(0, this._len).join(", ") + ">"
	}
}

